### PR TITLE
Feat/UI texture atlas

### LINKED ISSUE #1169: Using sprite atlas with the UI
**What problem does this solve or what need does it fill?**

I would like to be able to use something like `SpriteSheetBundle` but for the UI. I have my UI sprites in an atlas and it would be convenient to be able to use it and not have to export the individual images with `ImageBundle`

**Describe the solution would you like?**

To be able to use `SpriteSheetBundle` as a UI element - be able to style it and position it, etc.

**Describe the alternative(s) you've considered?**
**Additional context**

I did not know it is not possible and I was actually trying to do something like this:

```
    let player_texture_handle = asset_server.load("heart.png");
    let texture_atlas = TextureAtlas::from_grid(player_texture_handle, Vec2::new(16.0, 16.0), 2, 1);
    let texture_atlas_handle = texture_atlases.add(texture_atlas);

    commands.spawn(NodeBundle {
        style: Style {
            position_type: PositionType::Absolute,
            justify_content: JustifyContent::Center,
            align_items: AlignItems::FlexEnd,
            ..Default::default()
        },
        ..Default::default()
    })
    .with_children(|parent| {
        parent.spawn(SpriteSheetBundle {
            texture_atlas: texture_atlas_handle,
            transform: Transform::from_translation(Vec3::new(0.0, 0.0, 0.0)),
            ..Default::default()
        });
    });
```

And it did not work. I understand it is because I am mixing UI and non-UI elements but I think the error message should be more helpful, I was really struggling to understand what is wrong. The error I got:

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value'
```

So I think a bit nicer error message would be also welcome for every other beginner out there like I am :)

Thanks!

### PR DESCRIPTION
**Pull request by ManevilleF**
# Objective

Add support for texture sheets in `bevy_ui`. (Previous attempt: #3792).

Texture sheet support can enable:
- Animated buttons (or any UI element)
- Global texture mapping for UI

> Closes #1169

## Solution

I applied the exact same logic as #5103.
To use texture atlases for UI, adding a `TextureAtlas` component is enough to make it work.

## Note

This MR is now based on #5103 I will rebase this branch on main once the base MR is merged

### CODE REVIEW

File: crates/bevy_ui/src/ui_node.rs
Comment (mockersf) [inline crates/bevy_ui/src/ui_node.rs:386]: Having the texture index part of this variant of the enum would be better. It doesn't make sense to have a `UiTextureIndex` without a `UiImage::TextureAtlas`

Comment (ManevilleF) [reply crates/bevy_ui/src/ui_node.rs:387]: That's indeed better, I don't know why I didn't think of that :o

**Review by alice-i-cecile — COMMENTED** [review]
Interesting. From an ergonomics perspective I *really* like the enum approach. Especially once you migrate the index into the enum's data.

I actually think we should use this same strategy for 2D unless there are serious performance regressions; it makes the mental model and API much better. If we go that route, this should be done as two PRs:

1. Migrate 2D sprites to this enum-ful approach.
2. Add texture atlas support for UI.

@cart, do you have strong opinions here?

File: crates/bevy_ui/src/entity.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/entity.rs:99]: We should consider splitting this change out into its own PR and merging it ASAP.

File: crates/bevy_ui/src/render/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/render/mod.rs:320]: This change could also be split out.

File: crates/bevy_ui/src/entity.rs
Comment (ManevilleF) [reply crates/bevy_ui/src/entity.rs:99]: Okay, I'll remove it from this one

File: crates/bevy_ui/src/render/mod.rs
Comment (ManevilleF) [reply crates/bevy_ui/src/render/mod.rs:320]: I'll create a separate PR for this kind of improvements

**Comment by ManevilleF** [comment]
So everything seems to be working, I had some issues with the `Rect` computation on the extraction stage but it's all good now.

@alice-i-cecile About `bevy_sprite` and ergonomics overall I agree. I think this approach is cleaner and applying it to sprites would bring improvements like:
- No need for `SpriteSheetBundle`
- No need for `TextureAtlasSprite` which is very redundant with the `Sprite` component

If this approach is accepted I can open a different MR to bring these changes to `bevy_sprite`

File: crates/bevy_ui/src/entity.rs
Comment (ManevilleF) [reply crates/bevy_ui/src/entity.rs:139]: Damn

**Review by alice-i-cecile — APPROVED** [review]
Very nice. I'm happy with this now, and would love to see the rest of 2D unified under this approach. Adding this to Cart's Attention Board.

**Comment by ManevilleF** [comment]
> Very nice. I'm happy with this now, and would love to see the rest of 2D unified under this approach. Adding this to Cart's Attention Board.

Thanks, I opened a second PR for sprites here: #5072

**Comment by ManevilleF** [comment]
I rebased this MR upon #5103 in order to use the same logic and components, I will rebase it on `main` once #5103 is merged and only keep the UI related changes

**Comment by ManevilleF** [comment]
Rebased on #5103

**Comment by ManevilleF** [comment]
Rebased on #5103

**Comment by rparrett** [comment]
Could you mark this as closing #1169?

**Comment by lewiszlw** [comment]
Really useful feature, hope users can use it asap.

**Comment by nicoburns** [comment]
When using an image as a UI node, the image's intrinsic size is used as part of layout (if/when the layout engine requests that node's content-based size). TextureAtlas also seem to have an intrinsic size (actually two: the size in the atlas and an optional override size). Would it make sense to use those sizes for layout when TextureAtlas is used on a UI node?

(my intuition is that it would, but it probably doesn't need to block this PR and could be added later).

**Comment by alice-i-cecile** [comment]
Closing in favor of #8822, which isn't blocked on a refactor.

**Pull request closed without merging**
